File: backend/src/app/schemas/session_slot.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime
from typing import Optional, List

# Forward declaration for UserAvailability within SessionSlot
from .user_availability import UserAvailability

# Base schema with common attributes
class SessionSlotBase(BaseModel):
    slot_from: datetime
    slot_to: datetime
    note: Optional[str] = None

# Schema for creating a session slot (input)
# session_id will come from the path parameter
class SessionSlotCreate(SessionSlotBase):
    pass

# Schema for updating a session slot (input)
class SessionSlotUpdate(SessionSlotBase):
    # All fields are optional for update
    slot_from: Optional[datetime] = None
    slot_to: Optional[datetime] = None
    note: Optional[str] = None

# Schema for reading a session slot (output)
class SessionSlot(SessionSlotBase):
    id: int
    session_id: int
    created_at: datetime
    updated_at: datetime
    # Include user availabilities when reading a slot
    user_availabilities: List[UserAvailability] = []

    model_config = ConfigDict(from_attributes=True) # Replace orm_mode 

File: frontend/src/types/session_slot.ts
import type { UserAvailability } from './user_availability';

/**
 * Base interface for SessionSlot data (common fields).
 */
export interface SessionSlotBase {
  slot_from: string; // ISO 8601 datetime string
  slot_to: string;   // ISO 8601 datetime string
  note?: string | null;
}

/**
 * Interface for creating a new SessionSlot.
 */
export interface SessionSlotCreate extends SessionSlotBase {}

/**
 * Interface for updating an existing SessionSlot.
 * All fields are optional.
 */
export interface SessionSlotUpdate {
  slot_from?: string;
  slot_to?: string;
  note?: string | null;
}

/**
 * Interface representing a SessionSlot received from the API.
 */
export interface SessionSlot extends SessionSlotBase {
  id: number;
  session_id: number;
  created_at: string; // ISO 8601 datetime string
  updated_at: string; // ISO 8601 datetime string
  user_availabilities: UserAvailability[]; // Embed user availabilities
} 